### PR TITLE
[Core][VLM] Test registration for OOT multimodal models

### LINKED ISSUE #8667: [Feature]: support out tree multimodal models
### 🚀 The feature, motivation and pitch

vllm >= 0.6 doesn't support out tree defined multimodal models:
```
rank0]:   File "/usr/local/lib/python3.8/dist-packages/vllm/entrypoints/llm.py", line 177, in __init__
[rank0]:     self.llm_engine = LLMEngine.from_engine_args(
[rank0]:   File "/usr/local/lib/python3.8/dist-packages/vllm/engine/llm_engine.py", line 535, in from_engine_args
[rank0]:     engine_config = engine_args.create_engine_config()
[rank0]:   File "/usr/local/lib/python3.8/dist-packages/vllm/engine/arg_utils.py", line 792, in create_engine_config
[rank0]:     model_config = ModelConfig(
[rank0]:   File "/usr/local/lib/python3.8/dist-packages/vllm/config.py", line 230, in __init__
[rank0]:     self.multimodal_config = self._init_multimodal_config(
[rank0]:   File "/usr/local/lib/python3.8/dist-packages/vllm/config.py", line 251, in _init_multimodal_config
[rank0]:     raise ValueError(
[rank0]: ValueError: limit_mm_per_prompt is only supported for multimodal models.
```

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by ywang96**
This PR formally the support for OOT multimodal model registration. This was enabled alongside automatic detection of multimodal models in #7168, but has yet to be properly tested and documented.

FIX #8667 

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/models/__init__.py
Comment (youkaichao) [inline vllm/model_executor/models/__init__.py:208]: do you need an `else` branch here?

**Comment by youkaichao** [comment]
can you add some tests?

**Comment by ywang96** [comment]
> can you add some tests?

Yea, I made this PR so that the person can use it to test if it's working for him to unblock him

File: vllm/model_executor/models/__init__.py
Comment (ywang96) [reply vllm/model_executor/models/__init__.py:208]: No, we use `_OOT_MULTIMODAL_MODELS` to only check if the model is multimodal, and we still use `_OOT_MODELS` to check if the model is actually supported.

This follows the same relationship between `_MODELS` and `_MULTIMODAL_MODELS`.

File: tests/conftest.py
Comment (youkaichao) [inline tests/conftest.py:885]: need to have a new value for `_dummy_path` .

Comment (youkaichao) [reply tests/conftest.py:885]: e.g. adding `_dummy_phi3v_path = os.path.join(temp_dir, "dummy_phi3v")`

File: tests/models/test_oot_registration.py
Comment (youkaichao) [inline tests/models/test_oot_registration.py:45]: send some images for testing?

File: vllm/model_executor/models/__init__.py
Comment (Jack47) [inline vllm/model_executor/models/__init__.py:201]: how about judge model_cls to see whether it supports SupportsMultiModal instead of using new option?

Comment (ywang96) [reply vllm/model_executor/models/__init__.py:201]: I think this is something we tried previously but didn't work. See the comments here
https://github.com/vllm-project/vllm/blob/92ba7e7477619ec81464ccb64a17226f3d5047bb/vllm/model_executor/models/__init__.py#L208-L212

**Comment by ywang96** [comment]
There's some issue with the ordering of when "is_multimodal" needs to be checked and when the model actually gets registered to vLLM...I'm still investigating.

**Comment by ywang96** [comment]
This PR should be working now. I have verified `test_oot_registration.py` passed locally in my environment.

File: vllm/model_executor/models/__init__.py
Comment (DarkLight1337) [reply vllm/model_executor/models/__init__.py:201]: It should be ok to check the model class inside `ModelRegistry.register_model` itself because in that case we already have a direct reference to model_cls`.

The issue we had previously occurred because the vLLM model files may import other modules which initialize CUDA early. But in this case it is up to the user to not initialize CUDA early for their own models.

File: vllm/config.py
Comment (youkaichao) [inline vllm/config.py:157]: can you move it to `EngineArgs.__post_init__(self)` ? it looks strange if we directly specialize it for model config.

File: vllm/model_executor/models/__init__.py
Comment (ywang96) [reply vllm/model_executor/models/__init__.py:201]: I see what you mean, so we can infer whether a model is multimodal instead of asking the user to specify it via the flag.

Comment (DarkLight1337) [reply vllm/model_executor/models/__init__.py:200]: We have a dedicated function for this which is more robust than the issubclass check.

File: vllm/model_executor/models/__init__.py
Comment (DarkLight1337) [inline vllm/model_executor/models/__init__.py:206]:
```suggestion
        # NOTE: This is needed to store the information if the OOT model is
        # a multimodal model.
        if supports_multimodal(model_cls):
```

**Comment by ywang96** [comment]
Oops - meant to request instead of assigning, please take a look and let me know if there's any other issue @youkaichao @DarkLight1337

**Review by DarkLight1337 — APPROVED** [review]

**Review by youkaichao — APPROVED** [review]

**Comment by youkaichao** [comment]
> ValueError: The model's max seq len (131072) is larger than the maximum number of tokens that can be stored in KV cache (480). Try increasing `gpu_memory_utilization` or decreasing `max_model_len` when initializing the engine.

need to change some args to run the tests. @ywang96

**Comment by DarkLight1337** [comment]
Looks like CUDA reinitialization error is occurring because the model files are being imported by the plugins.

**Comment by DarkLight1337** [comment]
Imo the "right" way to fix this would be to update the model files themselves to avoid importing modules that initialize CUDA early, but this will probably require a lot of deferred imports which makes the code much harder to read.

**Comment by ywang96** [comment]
> Looks like CUDA reinitialization error is occurring because `phi3v.py` is being imported by the dummy plugin in `EngineArgs.__post_init__` which is before the vLLM engine is actually set up.

Hmmm.. I didn’t run into this issue when testing locally, but I can take a look today

**Comment by DarkLight1337** [comment]
I think basically all of the VLMs have this problem, because `xformers` is being imported at top level.

**Comment by DarkLight1337** [comment]
For now, I'm working around this by either running the test in a separate command, or using `VLLM_MULTIPROC_METHOD=spawn`. To solve this for real, we either have to:

- Update each model file so they don't do that at import time (which would require a lot of deferred imports)
- Make the OOT registration robust to early CUDA initialization (I don't think this is possible though. Correct me if I'm wrong @youkaichao)

**Comment by DarkLight1337** [comment]
Seems that now we have OOM error in entrypoints tests...

**Comment by ywang96** [comment]
> Seems that now we have OOM error in entrypoints tests...

Looks like it's actually caused by `test_audio.py`. I'm taking a look and see why it had connection error.

**Comment by DarkLight1337** [comment]
https://github.com/vllm-project/vllm/pull/8717/commits/52b600b97f36860b3ffbe49eeaca819610f4c8c2 seems to show that CUDA initialization in any of the plugins will interfere with vLLM's memory allocation even when the plugin isn't being used. This isn't good...

**Comment by ywang96** [comment]
@DarkLight1337 Turned out switching the order solved the issue...I'm honestly not sure why though :/

**Comment by DarkLight1337** [comment]
@youkaichao any idea about this?

**Comment by youkaichao** [comment]
sorry I missed the context here. can you summarize the current status? DM is also fine.

**Comment by ywang96** [comment]
> sorry I missed the context here. can you summarize the current status? DM is also fine.

Yea I shared with you on slack - I think it actually has nothing to do with CUDA re-init

**Comment by DarkLight1337** [comment]
Please don't merge this yet - we still need to investigate how installing the plugin is causing the test to fail.

**Comment by DarkLight1337** [comment]
I've added an option to lazy import the registered models, see if this fixes the problem.

**Comment by DarkLight1337** [comment]
Alright, tests pass now. @youkaichao feel free to merge if you are ok with the latest changes.

**Pull request merged**
